LG-140 Add ability to remember a phone for 2FA

jmhooper opened the pull request:
**Why**: So that users who log in repeatedly do not need to continually
2FA

**How**: Store an encrypted cookie that contains the user's ID and the
time the cookie was created. When the user goes to 2FA, check for that
cookie, and skip 2FA if it exists.

Hi! Before submitting your PR for review, and/or before merging it, please
go through the following checklist:

- [x] For DB changes, check for missing indexes, check to see if the changes
affect other apps (such as the dashboard), make sure the DB columns in the
various environments are properly populated, coordinate with devops, plan
migrations in separate steps.

- [x] For route changes, make sure GET requests don't change state or result in
destructive behavior. GET requests should only result in information being
read, not written.

- [x] For encryption changes, make sure it is compatible with data that was
encrypted with the old code.

- [x] For secrets changes, [make sure to update the S3 secrets bucket](https://github.com/18F/identity-private/wiki/Secrets-S3-buckets) with the 
new configs in **all** environments. 

- [x] Do not disable Rubocop or Reek offenses unless you are absolutely sure
they are false positives. If you're not sure how to fix the offense, please
ask a teammate.

- [x] When reading data, write tests for nil values, empty strings,
and invalid formats.

- [x] When calling `redirect_to` in a controller, use `_url`, not `_path`.

- [x] When adding user data to the session, use the `user_session` helper
instead of the `session` helper so the data does not persist beyond the user's
session.

- [x] When adding a new controller that requires the user to be fully
authenticated, make sure to add `before_action :confirm_two_factor_authenticated`.

jmhooper commented:
One thing to not here is that I've only added remember device to 2FA for phone. We'll need to add if for authenticator apps in a follow-on. That may be something to look into when we rethink how we handle the code for 2FA in order to add more 2FA methods.

jgsmith-usds commented on app/services/remember_device_cookie.rb:
```diff
@@ -0,0 +1,30 @@
+  def to_json
+    {
+      user_id: user_id,
+      created_at: created_at.iso8601,
```
I'd add some more information here (based in part on https://www.immun.io/blog/be-careful-signing/encrypting-rails-cookies). At least a sentinel like `role: :remember_me,` (and check that the unencrypted value has this), and some random data to increase the entropy of the encrypted values (makes cracking the encryption key a bit more difficult since we otherwise have fairly easily guessable information).

jmhooper replied:
Good call, I'll pull that together

jgsmith-usds replied:
LGTM

jgsmith-usds commented on app/services/remember_device_cookie.rb:
```diff
@@ -0,0 +1,30 @@
+  end
+
+  def self.from_json(json)
+    parsed_json = JSON.parse(json)
```
Here's where we'd reject this if the parsed JSON didn't have our sentinel indicating this is from a proper remember device cookie.

monfresh commented:
Any reason not to use the "remember me" functionality that's already built in to the `two_factor_authentication` gem? You can set the length of time with `config.remember_otp_session_for_seconds`.
https://github.com/Houdini/two_factor_authentication/blob/82fb6b192cb8c2ad98442410e934343dabf78118/app/controllers/devise/two_factor_authentication_controller.rb#L44-L53

jmhooper commented:
@monfresh: It looks like the gems implementation misses a few things we're going to want, like the ability to revoke the token if the user changes their phone number.

monfresh commented on app/views/two_factor_authentication/otp_verification/show.html.slim:
```diff
@@ -9,12 +9,16 @@ p == @presenter.phone_number_message
+  - if @presenter.remember_device_available?
+    .border.col-9.rounded-lg.mt2
+      = check_box_tag 'remember_device', true, false, class: 'my2 ml2 mr1'
+      = label_tag 'remember_device', 'Remember this device', class: 'blue'
```
Should this say how many days the device will be remembered?

jmhooper replied:
What do ya think @micahtaylor?

micahtaylor replied:
That would be ideal. The mock has it listed as `a month` - https://cm-jira.usa.gov/secure/attachment/55472/03e%20-%202fa%20-%20enter%20code%20%26%20remember%20device.png
Is that accurate?

jmhooper replied:
Yep, missed that. What do you think of "Remember this computer for n days"?

micahtaylor replied:
Is it `30 days`? If so, I'm fine just generalizing it to `Remember this computer for one month`

jmhooper replied:
The default is 30 days, but it is configurable, so we can change it without having to make a code change. It would be nice if changing the config to also update the text here. That's a bit easier if we use days.

If it is important, we can probably build a smarter helper that figures out what unit of time to use based on the number.

monfresh commented on spec/controllers/two_factor_authentication/otp_verification_controller_spec.rb:
```diff
@@ -359,6 +392,21 @@
       end
+
+      context 'when the user chooses to remember their device' do
+        it 'does not save a remember_device cookie' do
```
Is the title of this test right?

jmhooper replied:
Yeah, I think it could be named better perhaps. This is nested under the confirmation context.

One thing to check with @micahtaylor about is whether the "remember device" option should be available during device confirmation?

micahtaylor replied:
@andrewhughey and I went back and forth between that and landed on this page - https://cm-jira.usa.gov/secure/attachment/55472/03e%20-%202fa%20-%20enter%20code%20%26%20remember%20device.png

jmhooper replied:
@micahtaylor: Yeah, so that page is reused for sign in, sign up, and when the user changes their phone. My question is about which of those screens we want the option on, or do we want it on all of them?

micahtaylor replied:
Ah, right. At the very least sign in. Depending on if there is any major level of effort it would be nice to have it on all 3.

jmhooper replied:
Right now I only have it on sign in, but I can do all 3.

monfresh commented on spec/controllers/two_factor_authentication/otp_verification_controller_spec.rb:
```diff
@@ -474,6 +522,21 @@
         end
       end
+
+      context 'when the user chooses to remember their device' do
```
Does the UI allow the user to remember their device in the IdV context? If not, do we still need this test? If we just want to make sure someone can't make a POST request to remember a device (with cURL for example), maybe we should rename the test?

jmhooper replied:
:+1:

monfresh commented on spec/features/two_factor_authentication/remember_device_spec.rb:
```diff
@@ -0,0 +1,82 @@
+
+      # Sign in as second user and expect otp confirmation
+      sign_in_before_2fa(second_user)
+      # TODO: Figure out how to incrase 2FA limit
```
Can we remove this TODO now? Looks like you figured it out.

monfresh commented on spec/features/two_factor_authentication/remember_device_spec.rb:
```diff
@@ -0,0 +1,82 @@
+
+      first(:link, t('links.sign_out')).click
+
+      # Sign in as first user again and expect otp confirmation
```
Why does the first user see an OTP prompt when they sign back in?

jmhooper replied:
To keep the implementation simple, at first at least, I set it up so that it only remember one device at a time. That prevents a potential issue where multiple users in the cookie can cause it to grow too large and lead to issues decrypting or parsing it. We could work around that be adding a table to the database for remember device tokens, but that may be better as a follow on since it gets a bit complicated.

monfresh commented on app/controllers/concerns/remember_device_concern.rb:
```diff
@@ -0,0 +1,26 @@
+    return unless params[:remember_device] == 'true'
+    cookies.encrypted[:remember_device] = RememberDeviceCookie.new(
+      user_id: current_user.id,
+      created_at: Time.zone.now
```
Should we also set an expiration for the cookie equal to `Figaro.env.remember_device_expiration_days.to_i` from `Time.zone.now`?

jmhooper replied:
🤔, yeah good call

jmhooper commented:
@monfresh @jgsmith-usds: I've changed this one up a bit and would appreciate another look. I've added the remember device functionality to the update phone and sign up screens at the request of Micah. I've pulled together some shared examples to test remember device on all the screens where it shows up.

jgsmith-usds commented on config/locales/forms/fr.yml:
```diff
@@ -18,6 +18,7 @@ fr:
     messages:
       current_address: Vous devriez être en mesure de recevoir du courrier à cette
         adresse.
+      remember_device: NOT TRANSLATED YET
```
Do we need this translated or will this fall back to the English?

jmhooper replied:
This should fallback to english.

jgsmith-usds commented on app/services/remember_device_cookie.rb:
```diff
@@ -0,0 +1,51 @@
+  private_class_method def self.check_cookie_role(parsed_json)
+    role = parsed_json['role']
+    return if role == COOKIE_ROLE
+    raise "RememberDeviceCookie role '#{role}' did not match 'remember_me'"
```
Not a blocker for me, but if you're in editing, you might want to interpolate `COOKIE_ROLE` rather than hard-coding the value that it's not matching.

jmhooper replied:
Good catch, I agree

monfresh commented on config/application.yml.example:
```diff
@@ -128,6 +128,7 @@ development:
   reauthn_window: '120'
   redis_url: 'redis://localhost:6379/0'
   redis_throttle_url: 'redis://localhost:6379/1'
+  remember_device_expiration_days: '30'
```
Should we make this required in `Figaro.rb`? I ran this locally without updating my local application.yml and saw "Remember this device for days", which sounds hip, but not specific 😄

jmhooper replied:
Yeah, I'll add it

monfresh commented:
While testing this locally during first time sign up, I checked the "remember this device" checkbox, and then I clicked the link to get a phone call instead. The checkbox was no longer checked. Is there an easy way to remember the checked state? Is this worth doing?

monfresh commented:
I found a CSP bug in Chrome. Steps to reproduce:

1. Visit OIDC SP and click button to go to login.gov
2. During authentication or first time phone setup, check "remember this device", then sign in fully until you are redirected back to SP
3. Sign out from the SP
4. Repeat step 1
5. Sign in with email and password

Result: App stays stuck on the sign in screen. The following error shows up in the browser console:
```
Refused to send form data to 'http://localhost:9292/' because it violates the 
following Content Security Policy directive: "form-action 'self'".
```

jmhooper commented:
@monfresh:

RE remembering the selection: I think it is possible but is most likely worth doing in a follow on if we want it

RE the CSP issues: That's a really good catch. We'll need to the secure headers concern to the sign in screen.

jmhooper commented:
@monfresh: I've tied all of the above up. Any last thoughts on this one?

monfresh commented:
One thing that might potentially be confusing to users is what "device" refers to. I can imagine some users using shared computers might think "device" refers to their physical phone and might not realize that by checking the box, it will allow anyone else who has access to that computer to bypass 2FA on their account.

Would it make sense to add some text about shared computers?

andrewhughey commented:
@jmhooper I spoke with @mzia who said he had no specific concerns.
